ros/src/tl_detector/tl_detector.py

```diff
@@ -283,14 +295,31 @@ def process_traffic_lights(self):
                  (specified in styx_msgs/TrafficLight)
 
         """
+
+        # searching_distance_tl parameter sets distance in which
+        # traffic lights will be searched as number of waypoints
+        searching_distance_tl = 120
+
         light = None
-        light_positions = self.config['light_positions']
+        light_wp = -1
         if(self.pose and self.waypoints):
+            # In case that array with traffic light waypoints does not exist,
+            # create it
+            if(self.tl_waypoints == []):
+                self.get_tl_waypoints()
             car_position = self.get_closest_waypoint(self.pose.pose)
-
-        # TODO find the closest visible traffic light (if one exists)
-
-        if light:
+            # Loop thorugh waypoints to find the closest traffic ligt waypoint
+            smallest_tl_distance = 10000
+            for tl_waypoint in self.tl_waypoints:
+                distance_between_waypoints = abs(car_position-tl_waypoint)
```

Comment:
We definitely need a function for this, I think we have very similar code in 3 places now :)

Reply:
This function is not the same as get_closest_waypoint, here I am subtracting waypoints and not calculating distance between two coordinates.

Reply:
`get_closest_waypoint` is return the closest waypoint within `self.waypoints` to the input `Pose`. This piece of code is returning the closest `tl_waypoint` within `self.tl_waypoings` to the input `car_position`.

Maybe a generic function: `def get_closest_point(waypoints, pose)`? It's true that in this case you are using a few more conditions (like `smallest_tl_distance`) so it might be hard to refactor. But I think it would look nice (can be done in a separate patch)